### ISSUE TITLE
Audio poping/crackling on Raspberry Pi 4

### ISSUE THREAD

**Issue by fbarretto** [pygame/pygame#3571]
Hi,

I'm running a code that uses this soundController class that I've created with some .wav files. The games runs as a service in RPI4. The audio starts fine but at some point it start to get distorted and crackle. It happens only for the sounds that are played through pygame.  Any other audio played using aplay, youtube, etc is just fine. At the attached video you can see the tone played is glitching but the voice audio played with aplay is just fine.

Any hints of what might be going on?

```
import pygame
import glob
import os.path

class SoundController:
	
	def __init__(self, config):

		if config["main"]["environment"] == "test":
			self.basePath = config["path"]["base_dev"]
		else:
			self.basePath = config["path"]["base_deploy"]

		self.tonesPath = config["path"]["tones"]
		self.loopsPath = config["path"]["loops"]

		freq = 44100    # audio CD quality
		bitsize = -16   # unsigned 16 bit
		channels = 2    # 1 is mono, 2 is stereo
		buffer = 2048   # number of samples (experiment to get right sound)
		pygame.mixer.init(freq, bitsize, channels, buffer)
		pygame.mixer.music.set_volume(0.8)
		self.tones = []
		self.loadTones()
		self.loadLoops()

		#playTone(0)
	
	def playTone(self, toneId):
		self.tones[toneId].play()
	
	def playStart(self):
		self.gameStart.play()
	
	def playOver(self):
		self.gameOver.play()
	
	def loadLoops(self):
		self.gameOver = pygame.mixer.Sound(os.path.join(self.basePath, self.loopsPath, "error.wav"))
		self.gameStart = pygame.mixer.Sound(os.path.join(self.basePath, self.loopsPath, "start.wav"))
		
	def loadTones(self):
		files = glob.glob(os.path.join(self.basePath, self.tonesPath, '*.wav'))
		files.sort()
		for file in files:
			self.tones.append(pygame.mixer.Sound(file))

```

service
```
#memorize.service 
[Unit]
Description= Memorize Service
Wants=systemd-logind.service systemd-user-sessions.service display-manager.service
After=systemd-logind.service systemd-user-sessions.service display-manager.service

[Service]
Type = simple
Environment="DISPLAY=:0"
Environment="XAUTHORITY=/home/pi/.Xauthority"
ExecStartPre=/bin/sleep 60
ExecStart= /usr/bin/python3 /home/pi/memorize/main.py
User=pi
Restart=always
RestartSec=60
KillMode=process
SendSIGHUP=no

[Install]
WantedBy= graphical.target
```

I've found this issue that seems to be close to mine, but it has no solution.

https://user-images.githubusercontent.com/5246088/202352289-fd64ae8b-4f0e-4e0d-9dde-373c3856a98d.mp4

**Comment by yudonglin** [comment]
I encounter the same problem on pc with windows 11 and pygame 2.1.3.dev8 (SDL 2.24.0, Python 3.10.9). This similar to be a wav file related issue.

**Comment by yudonglin** [comment]
Is there any way that you can share your audio files? I think I am encountering a somewhat similar problem on pc with wav file. This could be a generalize wav file issue but with different behavior on different platform.

**Comment by fbarretto** [comment]
Sure, here is a sample of the files I'm using as wav, ogg and mp3. I cant recall exactly but I'm pretty sure I've tested with .ogg and mp3 files with the same outcome. 

I was launching the python game as a service (systemd) after boot and getting this issue, but running from command line didn't cause the issue so I've switched from service to autostart and that seems to do the trick. My guess is that something goes wrong when running pygame as service, possibly because of users/permissions but I didn't have enough time to investigate that.

[sound_samples.zip](https://github.com/pygame/pygame/files/10330070/sound_samples.zip)

**Comment by MyreMylar** [comment]
See:

https://raspberrypi.stackexchange.com/questions/132982/pygame-audio-is-full-of-pops-crackling-when-started-via-systemd

**Comment by sbuser** [comment]
Does this happen when you run the program directly (eg not as a service)? It does to me - when I play audio files via mixer I get periodic hisses and pops that I do not get when I play the same file through ffmpeg's ffplay.

In my case it has nothing to do with the service environment because I'm running the program directly and still get the behavior.

**Comment by yudonglin** [comment]
> 

I meet the same problem as you. @fbarretto 's audio files work just fine for me when I test them on my computer, but I assume this problem may have something to do with memory leaking somewhere. It will have a bigger impact on the Raspberry Pi since it has much less ram compared to a computer.

**Comment by sbuser** [comment]
In my case I believe I've isolated this to a raspberry pi issue with cpu priority or maybe with bus/irq activity.

Some detail: I'm using a pi camera which I'm continually capturing QR codes and also keyboard input in the background. When the python program is running (despite only using ~20% CPU) I get the hisses and pops in audio playback whether I play from external program like ffplay or via pygame's mixer.

So basically - if this is happening to you, try killing all other processes and using only a very simple pygame mixer or mixer.music to play the file. In my case - the files will play fine in that scenario.

**Comment by MyreMylar** [comment]
There isn't much we can do about the raspberry Pi's hardware limitations from pygame's side. 

Glad to hear you have made some progress in troubleshooting your issues though. I'll close this because there is no obvious work for pygame here, but people can still find it on google and add their thoughts.
